rfcs/999-kustomize.adoc

```diff
@@ -0,0 +1,265 @@
+  release: bikerental-deadbeef
+replicas:
+  - name: app
+    count: 12
+----
+
+== Artifact registry and Versioning
+
+Kustomizations could be easily versioned by hosting them in a tool such as
+GitHub or GitLab, as they can be referred to by a specific commit hash, branch,
+or tag. GitHub and GitLab make this easy by exposing URLs like
+`https://raw.githubusercontent.com/$ORG/$REPO/$REF/$PATH`.
+
+While pointing to a commit or tag solves versioning for specific versions of a
+kustomization, pointing to a branch comes with issues if want to support
+link:https://github.com/bookingcom/shipper/issues/62[version ranges]. There is
+no generic or standard way of resolving an URL pointing to a branch to a
+specific commit. This makes it impossible for Shipper to reliably roll back to
+an older kustomization within the specified range, as the URL would always
+resolve to the newest version.
+
+Our proposed solution to this is to host kustomizations in an OCI registry as a
+new artifact type. link:https://helm.sh/blog/helm-3-preview-pt3/[Helm 3] is
+taking that same route, so there is precedent to show we're not completely off
+the mark.
+
+By using an OCI registry, we get artifact tagging for free, and we also have
+the ability to resolve a tag to an immutable reference to an artifact. In this
+way, version ranges move away from Shipper, and into the artifact versioning
+scheme, giving kustomization authors more freedom in how to tag and update
+their artifacts.
+
+The actual process of using an OCI registry for kustomizations is outside the
+scope of this RFC, as
+link:https://stevelasker.blog/2019/05/11/authoring-oci-registry-artifacts-quick-guide/[there
+is good documentation for that out there].
+
+== Migration Path
+
+It is unreasonable to expect production users to instantly migrate away from
+Helm charts into Kustomize, so we also need a reasonable migration path. That
+means we'll need to support both Helm charts and Kustomize for a few versions.
+
+Besides the amount of code necessary to keep two application packaging running
+side by side, we do not foresee any major risks, provided that some simple
+rules to only ever use one or the other are followed.
+
+For this migration, the Application object will keep both `chart` and `values`
+alongside `kustomize`.  Our validating webhook will then ensure that either
+`chart` and `values` are populated, or only `kustomize`. At some point, after a
+few Shipper versions, the validating webhook can start rejecting objects with
+`chart` and `values`, while Shipper itself will still support both setups for a
+few more versions. Ultimately, Shipper will stop supporting Helm charts
+entirely, and any remaining applications using it will be invalid, and will no
+longer be managed by Shipper until they're fixed.
+
+Since we do Semantic Versioning, and to avoid the need of a new major version
+release when we do remove support for Helm charts, Helm support should be
+dropped before Shipper turns 1.0.
```

Comment:
Would be nice to have this migration plan as a series of steps instead of a paragraph to emphasize the scope of each step.